### PR TITLE
Append hosts to dependency container's /etc/hosts file

### PR DESCRIPTION
**Pull request by haircommander**
Before, any container with a netNS dependency simply used its dependency container's hosts file, and didn't abide its configuration (mainly --add-host). Fix this by always appending to the dependency container's hosts file, creating one if necessary.

fixes: #2504 

Note: there will be a different PR for a similar action with resolv.conf, but considering resolv.conf is more complex, and this fixes the issue, I decided to file this first.

### CODE REVIEW

File: libpod/container_internal_linux.go [hunk 1]
```diff
@@ -698,11 +702,37 @@ func (c *Container) makeBindMounts() error {
 			// If it doesn't, don't copy them
 			resolvPath, exists := bindMounts["/etc/resolv.conf"]
 			if exists {
-
 				c.state.BindMounts["/etc/resolv.conf"] = resolvPath
 			}
+
+			// check if dependency container has an /etc/hosts file
 			hostsPath, exists := bindMounts["/etc/hosts"]
```
Comment (TomSweeneyRedHat) [inline libpod/container_internal_linux.go]: You've a defaultHostsFile defined now, could you use it here?

Comment (haircommander) [reply libpod/container_internal_linux.go]: @TomSweeneyRedHat I wasn't sure if there should be a distinction of the map access in bind mounts (which happens quite a bit), and the access of the physical file /etc/hosts on the host system. If you think the former should be added, I'd also like to do the same with resolv.conf

File: libpod/container_internal_linux.go [hunk 2]
```diff
@@ -712,7 +742,7 @@ func (c *Container) makeBindMounts() error {
 			}
 			c.state.BindMounts["/etc/resolv.conf"] = newResolv
 
-			newHosts, err := c.generateHosts()
+			newHosts, err := c.generateHosts("/etc/hosts", c)
```
Comment (TomSweeneyRedHat) [inline libpod/container_internal_linux.go]: I'd use defaultHostFile here too

Comment (haircommander) [reply libpod/container_internal_linux.go]: aw shoot that was the point of adding it how did I miss it lol

File: libpod/container_internal_linux.go [hunk 3]
```diff
@@ -698,11 +702,37 @@ func (c *Container) makeBindMounts() error {
...
+				hostsPath, err = depCtr.generateHosts(hostsPath, c)
+
+				if err != nil {
+					depCtr.lock.Unlock()
+					return errors.Wrapf(err, "error creating hosts file for container %s which depends on container %s", c.ID(), depCtr.ID())
+				}
+				// If this file didn't exist before, save it in the depCtr's bind mount
+				// so it isn't overwritten in the future
+				if !exists {
+					depCtr.state.BindMounts["/etc/hosts"] = hostsPath
```
Comment (TomSweeneyRedHat) [inline libpod/container_internal_linux.go]: ditto defaultHostFile

File: libpod/container_internal_linux.go [hunk 4]
```diff
@@ -698,11 +702,37 @@ func (c *Container) makeBindMounts() error {
...
+			// check if dependency container has an /etc/hosts file
 			hostsPath, exists := bindMounts["/etc/hosts"]
-			if exists {
+
+			// if not, and the new container has hosts to add, set the hosts basis path to /etc/hosts
+			if !exists && (len(c.config.HostAdd) > 0 || (len(c.state.NetworkStatus) > 0 && len(c.state.NetworkStatus[0].IPs) > 0)) {
+				hostsPath = defaultHostFile
+			}
+			// only go through all of this if we want to add a hosts file
+			if len(hostsPath) > 0 {
```
Comment (TomSweeneyRedHat) [inline libpod/container_internal_linux.go]: I'll let others way in.  I'm muddling if we should remove the checks except " if !exists {" on line 712 or not?  I think if a container doesn't have a hosts file we'd want to create it regardless?  @mheon thoughts?

File: libpod/container_internal_linux.go [hunk 5]
```diff
@@ -698,11 +702,37 @@ func (c *Container) makeBindMounts() error {
...
+
+				if err != nil {
+					depCtr.lock.Unlock()
+					return errors.Wrapf(err, "error creating hosts file for container %s which depends on container %s", c.ID(), depCtr.ID())
+				}
+				// If this file didn't exist before, save it in the depCtr's bind mount
+				// so it isn't overwritten in the future
+				if !exists {
+					depCtr.state.BindMounts["/etc/hosts"] = hostsPath
+					depCtr.save()
```
Comment (mheon) [inline libpod/container_internal_linux.go]: Should catch this error and log it, if nothing else

File: libpod/container_internal_linux.go [hunk 6]
```diff
@@ -698,11 +702,37 @@ func (c *Container) makeBindMounts() error {
...
+			if !exists && (len(c.config.HostAdd) > 0 || (len(c.state.NetworkStatus) > 0 && len(c.state.NetworkStatus[0].IPs) > 0)) {
+				hostsPath = defaultHostFile
+			}
+			// only go through all of this if we want to add a hosts file
+			if len(hostsPath) > 0 {
+				depCtr.lock.Lock()
+				// generate a hosts file for the dependency container,
+				// based on either its old hosts file, or the default,
+				// and add the relevant information from the new container (hosts and IP)
+				hostsPath, err = depCtr.generateHosts(hostsPath, c)
```
Comment (mheon) [inline libpod/container_internal_linux.go]: I dunno if I like this... We won't pick this up until a restart, so it's not really shared.

Do we need to force sharing of /etc/hosts in this case, or is it sufficient to create it just for this container? We override resolv.conf for custom DNS settings when sharing network namespaces.

Comment (haircommander) [reply libpod/container_internal_linux.go]: How would other containers that are added to the namespace see it if a container they don't depend on (because they are laterally equal in the dependency tree)? 

My other question is: where besides /etc/host creation is the BindMount() map needed? If a dependency container was started up and had an empty /etc/hosts, but needed it for whatever process that was running (that shouldn't be restarted to see these changes), wouldn't that process see the changes when the new container appended to the depCtr's /etc/hosts file? I guess the question is: the state of a process that runs will be out of sync, but only on the libpod end and not from the perspective of namespace management, which I think would work, right?

**Comment by rhatdan** [comment]
Containers in the same Network Namespace are supposed to share the same /etc/resolv.conf and /etc/hosts file.  These files either need to be created when the pod is created or the first container is added to a pod. Then each container that gets added to the pod needs to share the pods (initial containers) network and thus its resolv.conf and hosts file.

Currently podman run -t --net=container:UUID 
Does the right thing for sharing hosts and resolv.conf.  This code has to be extended for pods.

Bottom line if there are 10 containers in a POD their is ONE resolv.conf and ONE hosts file bind mounted into each container.

File: libpod/container_internal_linux.go (hunk 6 above)
Comment (mheon) [reply libpod/container_internal_linux.go]: We use the map to determine what bind mounts to add when creating the container spec. If we add to it after container creation time, it will never be picked up before a restart. Actually, now that I think about it, we force a regeneration of the bind mounts on each restart, so we'll never pick up this change - we'll overwrite it on restart.

**Comment by mheon** [comment]
The issue is that we don't always create `/etc/hosts` as of now.

I think the solution is to force an unconditional creation/bind mount of `/etc/hosts` from `/etc/hosts` in the image. Potential problem: we won't commit changes to hosts when we commit the container anymore.

**Comment by rhatdan** [comment]
I think that is fine.  Does Docker commit changes to /etc/hosts?  I would figure that is usually a mistake to commit them.

**Comment by haircommander** [comment]
/etc/hosts was unconditionally created. This PR now errors out if for some reason it wasn't but the dependency container is one for netNS.

@rhatdan containers in the netNS did share the hosts file, the problem was --add-host didn't write anything to the dependency containers hosts, just ignored the flag. This PR now appends the contents of a new container using the --add-host flag, which fixes the issue.

PTAL @mheon @rhatdan

**Comment by haircommander** [comment]
Also @mheon and I talked about the restart problem (a dependency container won't retain the hosts a new container adds to it after restarting), and it was decided there wasn't anything to do about it.

**Comment by rhatdan** [comment]
@haircommander Lets talk about this tomorrow.

**Comment by haircommander** [comment]
bot, retest this please

File: libpod/container_internal_linux.go [hunk 7]
```diff
@@ -698,21 +698,37 @@ func (c *Container) makeBindMounts() error {
...
+
+			// check if dependency container has an /etc/hosts file
 			hostsPath, exists := bindMounts["/etc/hosts"]
-			if exists {
-				c.state.BindMounts["/etc/hosts"] = hostsPath
+			if !exists {
+				return errors.Errorf("error finding hosts file of dependency container %s for container %s", depCtr.ID(), c.ID())
+			}
+
+			depCtr.lock.Lock()
```
Comment (mheon) [inline libpod/container_internal_linux.go]: I think we can drop the lock here. We only really want to read the hosts file; If we fail to open it, we get a sane enough error back, if we succeed, we have a good file descriptor and can read the whole thing, even if the container is stopped

File: libpod/container_internal_linux.go [hunk 8]
```diff
@@ -698,21 +698,37 @@ func (c *Container) makeBindMounts() error {
...
+			// generate a hosts file for the dependency container,
+			// based on either its old hosts file, or the default,
+			// and add the relevant information from the new container (hosts and IP)
+			hostsPath, err = depCtr.generateHosts(hostsPath, c)
```
Comment (mheon) [inline libpod/container_internal_linux.go]: Hmm. I think this would replace the file, which might now work. Might be safer to open it for append and write the new entries?

Comment (mheon) [reply libpod/container_internal_linux.go]: Not work, even

Comment (haircommander) [reply libpod/container_internal_linux.go]: right now it's actually taking the old file, appending to it, and writing it all back (replacing it). I had it this way to have the actual generate code (make a new hosts file) and the append (when jumping into a netNS) the same function. Should I seperate?

File: libpod/container_internal_linux.go (hunk 7 above)
Comment (haircommander) [reply libpod/container_internal_linux.go]: I just figured generateHosts assumed a lock. What if two containers are added at the same time and they both try to append to the fd at the same time? odd case, but I think it's safer to have the lock. i'll leave it up to you.

Comment (mheon) [reply libpod/container_internal_linux.go]: Good point on the parallel writes - yeah, safest to leave it, then.

File: libpod/container_internal_linux.go (hunk 8 above)
Comment (mheon) [reply libpod/container_internal_linux.go]: `writeStringToRundir()` removes the file in question (if present) and recreates it, which is my concern. I think we need a separate append version, yeah.

Comment (haircommander) [reply libpod/container_internal_linux.go]: Fixed!

**Comment by giuseppe** [comment]
it seems to break the use case of: `podman pod stop -f $POD && podman pod start $POD` with:

```
ERRO[0000] "error starting some containers: container already exists" 
ERRO[0000] "error creating hosts file for container 288b0a539086c0789da33391971504a67ac172b9cab609ac264776145ad0bd06 which depends on container 2deda01da8ff8c1f7c896997707f2e5f958a60254c072f8bbc0b619c0dfee325: unable to open /var/run/containers/storage/overlay-containers/2deda01da8ff8c1f7c896997707f2e5f958a60254c072f8bbc0b619c0dfee325/userdata/hosts: open /var/run/containers/storage/overlay-containers/2deda01da8ff8c1f7c896997707f2e5f958a60254c072f8bbc0b619c0dfee325/userdata/hosts: no such file or directory" 
Error: error creating hosts file for container 36799180c4a76ab1bdd6e0e83d94a22c7c1fe077b3a0db98926654662826e541 which depends on container 2deda01da8ff8c1f7c896997707f2e5f958a60254c072f8bbc0b619c0dfee325: unable to open /var/run/containers/storage/overlay-containers/2deda01da8ff8c1f7c896997707f2e5f958a60254c072f8bbc0b619c0dfee325/userdata/hosts: open /var/run/containers/storage/overlay-containers/2deda01da8ff8c1f7c896997707f2e5f958a60254c072f8bbc0b619c0dfee325/userdata/hosts: no such file or directory
```

**Comment by mheon** [comment]
Confirmed, looking at it

**Comment by haircommander** [comment]
I actually got it @mheon

**Comment by haircommander** [comment]
I know what's wrong here

**Comment by mheon** [comment]
Alright I think I know what's going on here... Amazed it isn't happening for resolv too.

**Comment by haircommander** [comment]
it's from assuming we had an /etc/hosts file I think

**Comment by haircommander** [comment]
nevermind it looks like #2518 fixes. good find @mheon
